niceman/tests/utils.py

```diff
@@ -544,13 +544,29 @@ def newfunc(*args, **kwargs):
 
 
 def assert_is_subset_dict_recur(a, b):
-    """Asserts that 'a' is a subset of 'b' (recursively checks children dicts)
+    """Asserts that 'a' is a subset of 'b' (recursively checks dicts and lists)
     """
+    # For dictionaries recursively check children that are in a
     if isinstance(a, dict) and isinstance(b, dict):
         for key in a:
             if key not in b:
                 raise AssertionError("Key %s is missing" % key)
             assert_is_subset_dict_recur(a[key], b[key])
+    # For lists, recurse for every value a to make sure it is in b
+    # (note: two items in a may match the same item in b)
+    elif isinstance(a, list) and isinstance(b, list):
+        for a_val in a:
+            found = False
+            for b_val in b:
+                try:
+                    assert_is_subset_dict_recur(a_val, b_val)
+                    found = True
+                    break
+                except AssertionError:
+                    pass
+            if not found:
+                raise AssertionError("Array value %s is missing" % a_val)
```

Comment:
FYI, Python has 'else' for loop which is triggered if there were no 'break':
```
$> python /tmp/l.py
loop without break
1
2
else: finished, no break

loop with break
1

loop without looping
else: we do get here
2 14642.....................................:Thu 26 Oct 2017 09:44:34 AM EDT:.
hopa:~
$> cat /tmp/l.py

print("loop without break")
for i in [1,2]:
    print i
else:
    print "else: finished, no break"

print("\nloop with break")
for i in [1,2]:
    print i
    break
else:
    raise RuntimeError("must not get here since we break")

print("\nloop without looping")
for i in []:
    print i
else:
    print "else: we do get here"
```

so above could be
```python
   for b_val in b:
          try:
                 assert_is_subset_dict_recur(a_val, b_val)
                 break
          except AssertionError:
                 pass
     else:
          raise AssertionError("Array value %s is missing" % a_val)
```
right?

Reply:
now this assert functioning wouldn't also match its name.  I wondered if we should keep an old one as is and then add a new extended one? or parametrize it to do this `list` subset matching optionally?  in some functions (those which used it insofar I guess) we might like to have its more strict operation, just really matching subsets of dict key:value pairs while otherwise needing lists to be precisely the same

Reply:
BTW, lovely otherwise ;)

Reply:
suggested API
```python
assert_is_subset_recur(a, b, dicts=True, lists=False)
```

Reply:
Just a Python mindblow:  If we had it as a function
```
    from functools import partial
    if not any(map(partial(is_subset_dict_recur, a_val), b)):
         raise AssertionError("Array value %s is missing" % a_val)
```

Reply:
Addressed in 9a93c0c